File: clients/typescript/test/live-list.test.ts
import { describe, it, expect, vi } from "vitest";
import { LiveCollection } from "../src/live/live-collection.js";
import type { RealtimeCallback, RealtimeEvent } from "../src/realtime.js";

function fakeRealtime() {
  const subs = new Map<string, Set<RealtimeCallback>>();
  return {
    service: {
      async subscribe(topic: string, cb: RealtimeCallback) {
        let set = subs.get(topic);
        if (!set) subs.set(topic, (set = new Set()));
        set.add(cb);
        return () => set!.delete(cb);
      },
      unsubscribe(topic: string, cb?: RealtimeCallback) {
        if (cb) subs.get(topic)?.delete(cb);
        else subs.delete(topic);
      },
    },
    emit(topic: string, event: RealtimeEvent) {
      for (const cb of subs.get(topic) ?? []) cb(event);
    },
  };
}

describe("LiveCollection.getList -> LiveList", () => {
  it("seeds items ordered by sort and notifies observers on changes", async () => {
    const rt = fakeRealtime();
    const reader = {
      getOne: vi.fn(),
      getList: vi.fn(async () => ({
        items: [
          { id: "b", title: "B", rank: 2 },
          { id: "a", title: "A", rank: 1 },
        ],
        page: 1,
        perPage: 30,
        totalItems: 2,
      })),
      getPage: vi.fn(),
    };
    const lc = new LiveCollection("posts", reader as never, rt.service as never);

    const list = await lc.getList(1, 30, { sort: "rank" });
    expect(list.items.map((r) => r.id)).toEqual(["a", "b"]); // sorted by rank asc
    const cb = vi.fn();
    list.subscribe(cb);

    // matching create inserts at sorted position
    rt.emit("posts", {
      topic: "posts",
      action: "create",
      record: { id: "c", title: "C", rank: 0 },
    });
    expect(list.items.map((r) => r.id)).toEqual(["c", "a", "b"]);
    expect(cb).toHaveBeenCalled();
  });

  it("removes on delete and re-sorts in place when an update changes a sort key", async () => {
    const rt = fakeRealtime();
    const reader = {
      getOne: vi.fn(),
      getList: vi.fn(async () => ({
        items: [
          { id: "a", rank: 1 },
          { id: "b", rank: 2 },
          { id: "c", rank: 3 },
        ],
        page: 1,
        perPage: 30,
        totalItems: 3,
      })),
      getPage: vi.fn(),
    };
    const lc = new LiveCollection("posts", reader as never, rt.service as never);
    const list = await lc.getList(1, 30, { sort: "rank" });

    rt.emit("posts", { topic: "posts", action: "delete", record: { id: "b" } });
    expect(list.items.map((r) => r.id)).toEqual(["a", "c"]);

    // update moves "a" to the end and patches it in place
    rt.emit("posts", { topic: "posts", action: "update", record: { id: "a", rank: 9 } });
    expect(list.items.map((r) => r.id)).toEqual(["c", "a"]);
    expect((list.items[1] as unknown as { rank: number }).rank).toBe(9);
  });

  it("drops a record on an update that moves it OUT of an own-field filter", async () => {
    const rt = fakeRealtime();
    const reader = {
      getOne: vi.fn(),
      getList: vi.fn(async () => ({
        items: [
          { id: "a", status: "published", rank: 1 },
          { id: "b", status: "published", rank: 2 },
        ],
        page: 1,
        perPage: 30,
        totalItems: 2,
      })),
      getPage: vi.fn(),
    };
    const lc = new LiveCollection("posts", reader as never, rt.service as never);
    const list = await lc.getList(1, 30, { filter: "status = 'published'", sort: "rank" });

    rt.emit("posts", {
      topic: "posts",
      action: "update",
      record: { id: "a", status: "draft", rank: 1 },
    });
    expect(list.items.map((r) => r.id)).toEqual(["b"]);

    // an update that moves a NEW record INTO the filter inserts it
    rt.emit("posts", {
      topic: "posts",
      action: "update",
      record: { id: "z", status: "published", rank: 0 },
    });
    expect(list.items.map((r) => r.id)).toEqual(["z", "b"]);
  });
});

File: clients/dart/test/live/live_list_test.dart
import 'package:test/test.dart';
import 'package:zigbase_client/zigbase_client.dart';

import 'fakes.dart';

List<String> ids(LiveList list) => list.items.map((r) => r.id).toList();

void main() {
  group('LiveCollection.getList -> LiveList', () {
    test('seeds items ordered by sort and notifies on changes', () async {
      final rt = FakeLiveSubscriber();
      final reader = FakeReader()
        ..onGetListItems = () async => [
              rec({'id': 'b', 'title': 'B', 'rank': 2}),
              rec({'id': 'a', 'title': 'A', 'rank': 1}),
            ];
      final lc = LiveCollection('posts', reader, rt);

      final list = await lc.getList(sort: 'rank');
      expect(ids(list), ['a', 'b']); // sorted by rank asc
      var notifications = 0;
      list.changes.listen((_) => notifications += 1);

      rt.emit(
        'posts',
        RecordEvent(
            'posts', 'create', rec({'id': 'c', 'title': 'C', 'rank': 0})),
      );
      expect(ids(list), ['c', 'a', 'b']);
      await pumpEventQueue();
      expect(notifications, greaterThanOrEqualTo(1));
    });

    test('removes on delete and re-sorts in place when an update moves a key',
        () async {
      final rt = FakeLiveSubscriber();
      final reader = FakeReader()
        ..onGetListItems = () async => [
              rec({'id': 'a', 'rank': 1}),
              rec({'id': 'b', 'rank': 2}),
              rec({'id': 'c', 'rank': 3}),
            ];
      final lc = LiveCollection('posts', reader, rt);
      final list = await lc.getList(sort: 'rank');

      rt.emit('posts', RecordEvent('posts', 'delete', rec({'id': 'b'})));
      expect(ids(list), ['a', 'c']);

      // update moves "a" to the end and patches it in place
      rt.emit(
          'posts', RecordEvent('posts', 'update', rec({'id': 'a', 'rank': 9})));
      expect(ids(list), ['c', 'a']);
      expect(list.items[1]['rank'], 9);
    });

    test('drops a record on an update that moves it OUT of an own-field filter',
        () async {
      final rt = FakeLiveSubscriber();
      final reader = FakeReader()
        ..onGetListItems = () async => [
              rec({'id': 'a', 'status': 'published', 'rank': 1}),
              rec({'id': 'b', 'status': 'published', 'rank': 2}),
            ];
      final lc = LiveCollection('posts', reader, rt);
      final list =
          await lc.getList(filter: "status = 'published'", sort: 'rank');

      rt.emit(
        'posts',
        RecordEvent(
            'posts', 'update', rec({'id': 'a', 'status': 'draft', 'rank': 1})),
      );
      expect(ids(list), ['b']);

      // an update that moves a NEW record INTO the filter inserts it
      rt.emit(
        'posts',
        RecordEvent('posts', 'update',
            rec({'id': 'z', 'status': 'published', 'rank': 0})),
      );
      expect(ids(list), ['z', 'b']);
    });
  });

  group('LiveList sorted insertion', () {
    test('lands a matching create at the correct sorted position', () async {
      final rt = FakeLiveSubscriber();
      final reader = FakeReader()
        ..onGetListItems = () async => [
              rec({'id': 'a', 'rank': 10}),
              rec({'id': 'c', 'rank': 30}),
              rec({'id': 'e', 'rank': 50}),
            ];
      final lc = LiveCollection('posts', reader, rt);
      final list = await lc.getList(sort: 'rank');
      expect(ids(list), ['a', 'c', 'e']);

      rt.emit('posts',
          RecordEvent('posts', 'create', rec({'id': 'd', 'rank': 40})));
      expect(ids(list), ['a', 'c', 'd', 'e']); // middle

      rt.emit(
          'posts', RecordEvent('posts', 'create', rec({'id': 'z', 'rank': 5})));
      expect(ids(list), ['z', 'a', 'c', 'd', 'e']); // front

      rt.emit('posts',
          RecordEvent('posts', 'create', rec({'id': 'y', 'rank': 99})));
      expect(ids(list), ['z', 'a', 'c', 'd', 'e', 'y']); // end
    });

    test('O(1) membership: getById returns the live record', () async {
      final rt = FakeLiveSubscriber();
      final reader = FakeReader()
        ..onGetListItems = () async => [
              rec({'id': 'a', 'rank': 1})
            ];
      final lc = LiveCollection('posts', reader, rt);
      final list = await lc.getList(sort: 'rank');
      expect(list.getById('a')?.id, 'a');
      expect(list.getById('missing'), isNull);
    });

    test('id-asc tiebreaker orders records with equal sort keys', () async {
      final rt = FakeLiveSubscriber();
      final reader = FakeReader()
        ..onGetListItems = () async => [
              rec({'id': 'b', 'rank': 1}),
              rec({'id': 'a', 'rank': 1}),
            ];
      final lc = LiveCollection('posts', reader, rt);
      final list = await lc.getList(sort: 'rank');
      expect(ids(list), ['a', 'b']); // tie broken by id asc

      rt.emit(
          'posts', RecordEvent('posts', 'create', rec({'id': 'c', 'rank': 1})));
      expect(ids(list), ['a', 'b', 'c']);
    });
  });

  group('cache identity across views', () {
    test('the same id is one LiveRecord shared by a list and a getOne',
        () async {
      final rt = FakeLiveSubscriber();
      final reader = FakeReader();
      reader.onGetListItems = () async => [
            rec({'id': 'a', 'rank': 1})
          ];
      reader.onGetOne = (String id) async => rec({'id': id, 'rank': 1});
      final lc = LiveCollection('posts', reader, rt);
      final list = await lc.getList(sort: 'rank');
      final one = await lc.getOne('a');

      // Same backing record: an event patches both views at once.
      expect(one.get(), same(list.getById('a')!.get()));
      rt.emit(
          'posts', RecordEvent('posts', 'update', rec({'id': 'a', 'rank': 7})));
      expect(one['rank'], 7);
      expect(list.getById('a')!['rank'], 7);
    });
  });
}
